<!DOCTYPE html>
<html>
<head>
	<title>МЗКІТ Лабораторна робота № 15</title>
	<meta charset="utf-8">
	<link rel="stylesheet" type="text/css" href="style.css">

</head>
<body>
	<div class="paper">
		<div id="colontitle">МЗКІТ Лабораторна робота № 15</div>
		<ul class="lec-menu"> 
			<li><a href="index.html">Зміст</a></li> 
			<li><a href="lab-01.html">01</a></li> 
			<li><a href="lab-02.html">02</a></li> 
			<li><a href="lab-03.html">03</a></li> 
			<li><a href="lab-04.html">04</a></li> 
			<li><a href="lab-05.html">05</a></li> 
			<li><a href="lab-06.html">06</a></li> 
			<li><a href="lab-07.html">07</a></li> 
			<li><a href="lab-08.html">08</a></li> 
			<li><a href="lab-09.html">09</a></li> 
			<li><a href="lab-10.html">10</a></li>  
			<li><a href="lab-11.html">11</a></li> 
			<li><a href="lab-12.html">12</a></li> 
			<li><a href="lab-13.html">13</a></li> 
			<li><a href="lab-14.html">14</a></li> 
			<li><a href="lab-15.html">15</a></li> 
			<li><a href="lab-16.html">16</a></li> 
			<li><a href="lab-17.html">17</a></li> 
			<li><a href="lab-18.html">18</a></li> 
			<li><a href="lab-19.html">19</a></li> 
			<li><a href="lab-20.html">20</a></li>  
		</ul>  
		<h1>Обробка помилок обчислень</h1>
		
		<h2>Теоретичні відомості</h2>
		<h3>Обработка ошибок, "try..catch"</h3>
		<p>
			Неважно, насколько мы хороши в программировании, иногда наши скрипты содержат ошибки. Они могут возникать из-за наших промахов, неожиданного ввода пользователя, неправильного ответа сервера и по тысяче других причин.
		</p>
		<p>
			Обычно скрипт в случае ошибки «падает» (сразу же останавливается), с выводом ошибки в консоль.
		</p>
		<p>
			Но есть синтаксическая конструкция try..catch, которая позволяет «ловить» ошибки и вместо падения делать что-то более осмысленное.
		</p>

		<h3>Синтаксис «try…catch»</h3>
		<p>
			Конструкция try..catch состоит из двух основных блоков: try, и затем catch:
		</p>
		<pre>
try 
{
	// код...
} 
catch (err) 
{
	// обработка ошибки
}
		</pre>
		<p>
			Работает она так:
		</p>
		<p>
			Сначала выполняется код внутри блока try {...}.
			Если в нём нет ошибок, то блок catch(err) игнорируется: выполнение доходит до конца try и потом далее, полностью пропуская catch.
		</p>
		<p>
			Если же в нём возникает ошибка, то выполнение try прерывается, и поток управления переходит в начало catch(err). Переменная err (можно использовать любое имя) содержит объект ошибки с подробной информацией о произошедшем.
		</p>
		<figure>
			<img class="figc" src="img/try-catch-flow.svg">
			<figcaption>Приблизний вигляд меню</figcaption>
		</figure>
		<p>
			Таким образом, при ошибке в блоке try {…} скрипт не «падает», и мы получаем возможность обработать ошибку внутри catch.
		</p>

		<pre>
try 
{
	alert('Начало блока try');  // (1) <--
	lalala; // ошибка, переменная не определена!
	alert('Конец блока try (никогда не выполнится)');  // (2)
} 
catch(err) 
{
	alert(`Возникла ошибка!`); // (3) <--
}
		</pre>
		<p class="notice">
			try..catch работает только для ошибок, возникающих во время выполнения кода. Чтобы try..catch работал, код должен быть выполнимым. Он не сработает, если код синтаксически неверен, например, содержит несовпадающее количество фигурных скобок или другие ошибки.
		</p>

		<p>
			try..catch работает синхронно: исключение, которое произойдёт в коде, запланированном «на будущее», например в setTimeout, try..catch не поймает.
		</p>
		<h3>Объект ошибки</h3>
		<p>
			Когда возникает ошибка, JavaScript генерирует объект, содержащий её детали. Затем этот объект передаётся как аргумент в блок catch:
		</p>
		<pre>
try 
{
	// ...
} 
catch(err) 
{ // <-- объект ошибки, можно использовать другое название вместо err
	// ...
}
		</pre>
		<p>
			Для всех встроенных ошибок этот объект имеет два основных свойства:
		</p>
		<ul>
			<li>name - Имя ошибки. Например, для неопределённой переменной это "ReferenceError".</li>
			<li>message - Текстовое сообщение о деталях ошибки.</li>
			<li>stack - Текущий стек вызова: строка, содержащая информацию о последовательности вложенных вызовов, которые привели к ошибке. Используется в целях отладки.</li>
		</ul>

		<h2>Хід роботи</h2>	
		<ol>
			<li>Перевірити чи встановлено пакет програм web-розробника XAMPP</li>
			<li>Викликати панель керування xampp-control.exe</li>
			<li>Впевнитись, що web-сервер Apache запущений</li>
			<li>Перейти за адресою <code>http://127.0.0.1/</code> або <code>http://localhost/</code> та впевнитись, що сторінка вітання XAMPP завантажилась</li>
			<li>Очистити зміст каталогу <code>C:\xampp\htdocs\</code></li>
			<li>Створити файл <code>index.html</code> в середині каталогу <code>C:\xampp\htdocs\</code></li>
			<li>Перейти за адресою <code>http://127.0.0.1/</code> або <code>http://localhost/</code> та впевнитись, що сторінка, яку ви створили, завантажилася корректно</li>
			<li>Створити файл з назвою <code>script.js</code> та помістити його поряд з файлом <code>index.html</code></li>
			<li>Підключити файл <code>script.js</code> в секції <code>&lt;head&gt;</code></li>
			<li>За допомогою оператору <code>throw</code> згенерувати помилку та вивести повідомлення в консоль браузеру</li>
			<li>За допомогою оператору <code>throw</code> згенерувати помилку та вивести повідомлення за допомогою функції <code>alert();</code></li>
			<li>За допомогою оператору <code>throw</code> згенерувати помилку та вивести повідомлення у спливаючому вікні з поясненням помилки. Вікно реалізувати за допомогою прихованого блоку <code>&lt;div&gt;</code> керування вікном має бути реалізоване за допомогою JavaScript</li>
			<li>Імітувати помилку звернення до елементу сторінки якого не існує. Повідомлення про помилку реалізувати як у попередньому пункті</li>
			<li>Імітувати помилку виклику функції, якої не вснує. Повідомлення про помилку реалізувати як у попередньому пункті</li>
			<li>Код JavaScript розмістити у файлі <code>script.js</code> почергово та перевірити його роботу</li>
			<li>У "підвал" сторінки та файл <code>script.js</code> помістити інформацію про виконавця роботи: група, ПІБ, дата виконання.</li>
			<li>Для кожного етапу роботи зробити знімки екрану або скопіювати текст консолі та додати їх у звіт з описом кожного скіншота</li>
			<li>Зберегти звіт у форматі PDF</li>
			<li>Роздрукувати звіт та письмово відповісти на контрольні запитання</li>
		</ol>
		<h2>Контрольні питання</h2>
		<ol>
			<li>?</li>
			<li>?</li>
			<li>?</li>
			<li>?</li>
			<li>?</li>

		</ol>
		<h2>Довідники та додаткові матеріали</h2>
		<ol>
			<li id="link01"><a href="https://validator.w3.org/">Валідатор W3C</a></li>
			<li id="link02"><a href="https://learn.javascript.ru/">Современный учебник JavaScript</a></li>
			<li id="link03"><a href="http://htmlbook.ru/">Справочник по HTML</a></li>
		</ol>	
	</div>
</body>
</html>